core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1367,9 +1368,108 @@ public AlluxioURI translateUri(String uriStr) throws InvalidPathException {
+      if (e instanceof InvalidPathException) {
+        // Don't throw as this could be thrown due to ufs having been mounted concurrently.
+      } else {
+        throw new InvalidPathException(
+            String.format("Failed to translate and auto-mount path: %s", uriStr), e);
+      }
+    }
+    // Try translating the URI after auto-mount.
+    return mMountTable.translate(uriStr);
+  }
+
+  /**
+   * Attempts to mount given UFS Uri into designated auto-mount directory.
+   *
+   * Mount attempt will start from the mounting lowest path component, and go to higher path
+   * components if mounting fails due to access control violation.
+   *
+   * TODO(ggezer) Have fine-grained locking on mount paths instead of marking synchronized.
+   *
+   * @param ufsUri ufs Uri for auto-mount
+   * @throws AlluxioException
+   * @throws IOException
+   */
+  private synchronized void autoMountUfsForUri(AlluxioURI ufsUri)
+      throws AlluxioException, IOException {
+    // Generate lowest Alluxio mount root for given path.
+    String alluxioMountRoot =
+        PathUtils.concatPath(ServerConfiguration.get(PropertyKey.MASTER_AUTO_MOUNT_ROOT),
+            ufsUri.getScheme(), ufsUri.getAuthority());
+    // Read default options for auto-mounted UFSes.
+    boolean mountReadonly = ServerConfiguration.getBoolean(PropertyKey.MASTER_AUTO_MOUNT_READONLY);
+    boolean mountShared = ServerConfiguration.getBoolean(PropertyKey.MASTER_AUTO_MOUNT_SHARED);
+    // Try mounting UFS to Alluxio starting from its lowest path.
+    int pathComponentIndex = 0;
+    String currentPathComponent;
+    // Used to keep track of first directory that is created for this auto-mount call.
+    // It'll be deleted recursively if this mount fails.
+    AlluxioURI firstCreatedParent = null;
+    try {
+      while ((currentPathComponent = ufsUri.getLeadingPath(pathComponentIndex++)) != null) {
+        // Generate current Alluxio mount root.
+        AlluxioURI currentAlluxioRootUri =
+            new AlluxioURI(PathUtils.concatPath(alluxioMountRoot, currentPathComponent));
+        // Generate current UFS mount root.
+        AlluxioURI currentUfsRootUri =
+            new AlluxioURI(ufsUri.getScheme(), ufsUri.getAuthority(), currentPathComponent);
+        try {
+          // Ensure Alluxio parent path exist before mounting.
+          AlluxioURI currentRootParentUri = currentAlluxioRootUri.getParent();
+          if (!mInodeTree.inodePathExists(currentRootParentUri)) {
+            if (firstCreatedParent == null) {
+              // Track first created dir while preparing to mount.
+              firstCreatedParent = currentRootParentUri;
+            }
+            createDirectory(currentAlluxioRootUri.getParent(), CreateDirectoryContext
+                .mergeFrom(CreateDirectoryPOptions.newBuilder().setRecursive(true)));
+          }
+          // Mount current lowest path of UFS to Alluxio.
+          mount(currentAlluxioRootUri, currentUfsRootUri, MountContext.mergeFrom(
```

Comment:
Are we "probing" the mount command? Why not "probe" something like ufs.exists() or something?

Reply:
What should we be probing?